### PR TITLE
Remove legacy providers

### PR DESCRIPTION
**Pull request by comius**
## Motivation for features / changes
Legacy struct providers have been deprecated by Bazel. Replacing them with modern providers, will make it possible to simplify and remove legacy handling from Bazel.

Prerequiste: https://github.com/bazelbuild/rules_closure/pull/599 is merged and released.

Googlers, see cl/597800285.

### CODE REVIEW

File: tensorboard/defs/web.bzl [hunk 1]
```diff
@@ -14,259 +14,295 @@
...
+        label = str(ctx.label),
+        bind = "localhost:6006",
+        manifest = [long_path(ctx, man) for man in devserver_manifests.to_list()],
+        external_asset = [
+            struct(webpath = k, path = v)
+            for k, v in ctx.attr.external_assets.items()
+        ],
+    )
+    params_file = _new_file(ctx, "-params.pbtxt")
+    ctx.actions.write(output = params_file, content = proto.encode_text(params))
```
Comment (arcra) [inline tensorboard/defs/web.bzl:133]: This block (lines 123-141) seems to be repeated right above (lines 104-122).

AFAICT, the only difference is the way this method is called: `proto.encode_text(params)` vs `params.to_proto()`... probably a bad sync?

File: tensorboard/defs/web.bzl [hunk 2]
```diff
@@ -14,259 +14,295 @@
...
-              collect_runfiles(ctx.attr.data),
-          ])))
+    # Export data to parent rules. This uses the legacy, string-based
+    # provider mechanism for compatibility with the base `web_library`
+    # rule from rules_closure: because `tf_web_library`s may depend on
+    # either other `tf_web_library`s or base `web_library`s, the
+    # interfaces ~must be the same.
+    return [
+        WebFilesInfo(
+            manifest = manifest,
```
Comment (arcra) [inline tensorboard/defs/web.bzl:150]: Just checking, are both of these expected? manifest and manifests (one plural, one singular)?

File: tensorboard/defs/web.bzl [hunk 3]
```diff
@@ -14,259 +14,295 @@
...
+            ctx.label,
+            webpath,
+        ))
+    if webpath in webpaths.to_list():
+        _fail(ctx, "webpath %s was defined by %s when already defined by deps" % (
+            webpath,
+            ctx.label,
+        ))
+    new_webpaths.append(webpath)
+    manifest_srcs.append(struct(
```
Comment (arcra) [inline tensorboard/defs/web.bzl:248]: Just checking... I understand this change is removing use of `struct` (is that right?)

Should I be looking to catch any uses of struct? Or it it specific ones from how they're used by rules_closure, so it's tricky to know which ones are right and which ones aren't?

File: tensorboard/defs/web.bzl [hunk 4]
```diff
@@ -14,259 +14,295 @@
...
-      strip = strip[1:]
-    if strip.endswith("/"):
-      _fail(ctx, "strip_prefix should not end with /")
-    else:
-      strip += "/"
-  return strip
+    strip = ctx.attr.strip_prefix
+    if strip:
+        if strip.startswith("/"):
+            _fail(ctx, "strip_prefix should not end with /")
```
Comment (arcra) [inline tensorboard/defs/web.bzl:279]: Pre-existing issue, but this should be "should not start with /"?

**Review by arcra — CHANGES_REQUESTED** [review]

File: WORKSPACE [hunk 5]
```diff
@@ -191,6 +188,11 @@ http_archive(
     ],
 )
 
+# WORKAROUND for rules_webtesting not declaring used com_github_gorilla_mux repo:
```
[the remark this answers is not in the document]
Comment (comius) [reply WORKSPACE:191]: No, I tried. com_github_gorilla_mux needs setup for Go repositories. It doesn't work even when used after rules_closure.

File: tensorboard/defs/web.bzl (hunk 1 above)
Comment (comius) [reply tensorboard/defs/web.bzl]: Fixed.

File: tensorboard/defs/web.bzl (hunk 2 above)
Comment (comius) [reply tensorboard/defs/web.bzl:131]: This is what's in your code at the moment: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/defs/web.bzl#L126-L127

File: tensorboard/defs/web.bzl (hunk 3 above)
Comment (comius) [reply tensorboard/defs/web.bzl:229]: No, it's not removing all `structs`, just the structs returned directly from a rule. This one is not a problem.

File: tensorboard/defs/web.bzl (hunk 4 above)
Comment (comius) [reply tensorboard/defs/web.bzl:260]: I don't want to mix semantic changes into a cleanup cl. Please do the fixes as a followup.

**Review by arcra — APPROVED** [review]

**Pull request merged**